components/lark-pretty-print/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+use lark_entity::{Entity, EntityData, ItemKind, LangItem, MemberKind};
+use lark_intern::Untern;
+use lark_parser::ParserDatabase;
+use lark_ty::declaration::{Declaration, DeclaredPermKind};
+use lark_ty::full_inferred::{FullInferred, FullInferredTables};
+use lark_ty::{BaseData, BaseKind, BoundVarOr, PermKind, Ty, TypeFamily};
+
+pub trait PrettyPrintDatabase: ParserDatabase + AsRef<FullInferredTables> {}
+
+pub trait PrettyPrint {
+    fn pretty_print(&self, db: &impl PrettyPrintDatabase) -> String;
+}
+
+impl PrettyPrint for Ty<Declaration> {
+    fn pretty_print(&self, db: &impl PrettyPrintDatabase) -> String {
+        format!(
+            "{}{}",
+            match self.perm.untern(db) {
+                DeclaredPermKind::Own => "",
+            },
+            match self.base.untern(db) {
+                BoundVarOr::BoundVar(var) => format!("{:?}", var),
+                BoundVarOr::Known(base_data) => base_data.pretty_print(db),
+            }
+        )
+    }
+}
+
+impl PrettyPrint for Ty<FullInferred> {
+    fn pretty_print(&self, db: &impl PrettyPrintDatabase) -> String {
+        format!(
+            "{}{}",
+            match self.perm {
+                PermKind::Own => "",
+                PermKind::Share => "shared ",
+                PermKind::Borrow => "borrowed ",
+            },
+            self.base.untern(db).pretty_print(db),
+        )
+    }
+}
+
+impl<T: TypeFamily> PrettyPrint for BaseData<T> {
+    fn pretty_print(&self, db: &impl PrettyPrintDatabase) -> String {
+        self.kind.pretty_print(db)
+    }
+}
+
+impl PrettyPrint for Entity {
```

Comment:
This doesn't pretty print quite what I expected. I anticipated that it would pretty-print the *path* to an entity (so e.g. pretty printing a function `foo` might generate the path to refer to the function `foo`). But what it actually seems to do is to pretty print sort of depends on the type of the entity -- for a struct, it's the path, for a function, it's the signature, etc. Makes me wonder if we want to break out those different things into different helpers so we can be a bit clearer about what we want.